haskell/cabal.bzl

```diff
@@ -2626,3 +2634,54 @@ def _expand_make_variables(name, ctx, strings):
         ctx.attr.tools,
     ]
     return expand_make_variables(name, ctx, strings, extra_label_attrs)
+
+def _use_stack_impl(repository_ctx):
+    # sanity check
+    stack = repository_ctx.path(repository_ctx.attr.stack)
+    valid, version = _stack_version_check(repository_ctx, stack)
+    if not valid:
+        fail("Stack {} found. Need version {}.{} or newer.".format(version, *_STACK_MIN_VERSION))
+    repository_ctx.symlink(stack, "stack")
+
+_use_stack = repository_rule(
+    _use_stack_impl,
+    attrs = {
+        "stack": attr.label(),
+    },
+)
+
+def use_stack(stack):
+    """
+    use given `stack` binary in `stack_snapshot`.
+    allows overriding `stack` installation independent of call site,
+    specifically from repositories outside of where a `stack_snapshot`
+    has been declared.
+    call this in `WORKSPACE` before `stack_snapshot`.
+
+    Example:
+    # A/WORKSPACE
+    stack_snapshot(
+        name = "a",
+        stack = "@some_stack:stack",
+        # ...
+    )
+
+    # B/WORKSPACE
+    # order is important!
+    local_repository(
+        name = "my_stack",
+        path = "../some_custom_stack",
+    )
+    use_stack("@my_stack//:stack")
+    local_repository(
+        name = "A",
+        path = "../A",
+    )
+    # using `@A//:a` as dependency will build on `my_stack`
```

Comment:
That's not how Bazel workspaces work. Bazel will only consider *one* `WORKSPACE` file, the one in the main repository, all other `WORKSPACE` files are ignored.

So, the correct usage is to place `use_stack` into the `WORKSPACE` before any `stack_snapshot`.